File: dehaze-go/internal/api/client_log.go
package api

import (
	"strings"

	"github.com/earthyzinc/dehaze-go/internal/model/dto"
	"github.com/earthyzinc/dehaze-go/pkg/common"
	"github.com/earthyzinc/dehaze-go/pkg/logger"
	"github.com/earthyzinc/dehaze-go/pkg/security"
	"github.com/gin-gonic/gin"
	"go.uber.org/zap"
)

const (
	maxMessageLength = 2000
	maxErrorStackLen = 8000
)

type ClientLogApi struct{}

func NewClientLogApi() *ClientLogApi {
	return &ClientLogApi{}
}

// Collect 接收前端 SDK 批量上报的日志并落盘 client.log。
//
// 匿名（未登录）仅允许上报 ERROR 且必须携带 trace_id；已登录用户从会话解析 user_id 注入。
func (a *ClientLogApi) Collect(c *gin.Context) {
	var batch dto.ClientLogBatch
	if err := c.ShouldBind(&batch); err != nil {
		_ = c.Error(err)
		return
	}

	// security.GetUserID 未登录返回 0
	userID := security.GetUserID(c)
	clientLogger := logger.GetClientLogger()
	for _, entry := range batch.Logs {
		writeEntry(&entry, userID, clientLogger)
	}

	common.Ok(c)
}

// writeEntry 处理单条日志条目：匿名过滤、字段构建、按级别落盘。
// 拆分为接收 logger 参数，便于单元测试注入 observer logger。
func writeEntry(entry *dto.ClientLogEntry, userID int64, log *zap.Logger) {
	traceID := strings.TrimSpace(entry.TraceID)
	// 匿名仅允许上报 ERROR 且必须携带 trace_id，否则丢弃该条，避免被滥用刷日志
	if userID == 0 && (!isError(entry.Level) || traceID == "") {
		return
	}

	fields := buildFields(entry, traceID)
	if userID > 0 {
		fields = append(fields, zap.Int64("user_id", userID))
	}

	message := truncate(entry.Message, maxMessageLength)
	level := strings.ToUpper(strings.TrimSpace(entry.Level))
	switch level {
	case "ERROR":
		log.Error(message, fields...)
	case "WARN":
		log.Warn(message, fields...)
	default:
		log.Info(message, fields...)
	}
}

func buildFields(entry *dto.ClientLogEntry, traceID string) []zap.Field {
	fields := make([]zap.Field, 0, 20)
	// 与 Java 端 CharSequenceUtil.isNotBlank 对齐：过滤 null/空字符串/纯空白
	putIfNotBlank := func(key, value string) {
		if strings.TrimSpace(value) != "" {
			fields = append(fields, zap.String(key, value))
		}
	}

	// 不注入前端 timestamp：FileEncoder 已输出服务端接收时间的 timestamp，避免 JSON 同键冲突
	putIfNotBlank("app", entry.App)
	putIfNotBlank("app_version", entry.AppVersion)
	putIfNotBlank("url", entry.URL)
	putIfNotBlank("user_agent", entry.UserAgent)
	putIfNotBlank("error_type", entry.ErrorType)
	putIfNotBlank("error_source", entry.ErrorSource)
	putIfNotBlank("error_stack", truncate(entry.ErrorStack, maxErrorStackLen))
	putIfNotBlank("method", entry.Method)
	putIfNotBlank("path", entry.Path)
	putIfNotBlank("code", entry.Code)
	putIfNotBlank("type", entry.Type)
	putIfNotBlank("metric_name", entry.MetricName)
	putIfNotBlank("navigation_type", entry.NavigationType)
	putIfNotBlank("resource_url", entry.ResourceURL)
	putIfNotBlank("trace_id", traceID)
	if entry.Status != nil {
		fields = append(fields, zap.Int("status", *entry.Status))
	}
	if entry.Duration != nil {
		fields = append(fields, zap.Float64("duration", *entry.Duration))
	}
	if entry.MetricValue != nil {
		fields = append(fields, zap.Float64("metric_value", *entry.MetricValue))
	}
	return fields
}

func isError(level string) bool {
	return strings.EqualFold(strings.TrimSpace(level), "ERROR")
}

func truncate(value string, maxLength int) string {
	if len(value) > maxLength {
		return value[:maxLength]
	}
	return value
}

File: dehaze-python/app/router/client_log.py
"""前端日志接收路由。

接收前端 SDK 批量上报的日志，复用全局 RateLimitMiddleware（按 path+ip 限流），
已登录用户从会话解析 user_id 注入，未登录允许匿名上报（仅 ERROR 且必须带 trace_id），
以 NDJSON 形式写入 logs/{yyyy-MM-dd}/client.log 供 filebeat 采集。

字段规范见 dehaze-doc/docs/02-系统架构/07-日志架构设计.md §3.5。
"""

import logging
from typing import Optional

from fastapi import APIRouter, Body, Depends
from pydantic import BaseModel, Field

from app.core.code import ResultCode
from app.core.result import error, success
from app.dependencies.auth import UserContext, get_current_user_optional
from app.infrastructure.logging import get_client_logger

router = APIRouter(prefix="/api/v1/logs", tags=["前端日志"])

MAX_BATCH_SIZE = 50
MAX_MESSAGE_LENGTH = 2000
MAX_ERROR_STACK_LENGTH = 8000

logger = logging.getLogger(__name__)


class ClientLogEntry(BaseModel):
    """前端单条日志条目（见 07-日志架构设计.md §3.5.2）。"""

    timestamp: Optional[str] = None
    level: Optional[str] = None
    message: Optional[str] = None
    app: Optional[str] = None
    app_version: Optional[str] = None
    url: Optional[str] = None
    user_agent: Optional[str] = None
    trace_id: Optional[str] = None
    error_type: Optional[str] = None
    error_source: Optional[str] = None
    error_stack: Optional[str] = None
    method: Optional[str] = None
    path: Optional[str] = None
    status: Optional[int] = None
    duration: Optional[float] = None
    code: Optional[str] = None
    type: Optional[str] = None
    metric_name: Optional[str] = None
    metric_value: Optional[float] = None
    navigation_type: Optional[str] = None
    resource_url: Optional[str] = None


class ClientLogBatch(BaseModel):
    """前端日志批量上报请求体（单次最多 50 条）。"""

    logs: list[ClientLogEntry] = Field(..., min_length=1, max_length=MAX_BATCH_SIZE)


def _truncate(value: Optional[str], max_length: int) -> Optional[str]:
    if value is None:
        return None
    return value if len(value) <= max_length else value[:max_length]


def _is_error(level: Optional[str]) -> bool:
    return (level or "").upper() == "ERROR"


def _write_entry(entry: ClientLogEntry, user_id: Optional[int]) -> None:
    trace_id = (entry.trace_id or "").strip()

    # 匿名仅允许上报 ERROR 且必须携带 trace_id，否则丢弃该条，避免被滥用刷日志
    if user_id is None and (not _is_error(entry.level) or not trace_id):
        return

    fields = {
        key: value
        for key, value in {
            # 不注入前端 timestamp：ClientLogFormatter 已输出服务端接收时间的 timestamp，避免同键冲突
            "app": entry.app,
            "app_version": entry.app_version,
            "url": entry.url,
            "user_agent": entry.user_agent,
            "error_type": entry.error_type,
            "error_source": entry.error_source,
            "error_stack": _truncate(entry.error_stack, MAX_ERROR_STACK_LENGTH),
            "method": entry.method,
            "path": entry.path,
            "code": entry.code,
            "type": entry.type,
            "metric_name": entry.metric_name,
            "navigation_type": entry.navigation_type,
            "resource_url": entry.resource_url,
            "trace_id": trace_id,
        }.items()
        # 与 Java isNotBlank / Go TrimSpace 对齐：过滤 None 和纯空白字符串
        if value is not None and (not isinstance(value, str) or value.strip() != "")
    }
    if entry.status is not None:
        fields["status"] = entry.status
    if entry.duration is not None:
        fields["duration"] = entry.duration
    if entry.metric_value is not None:
        fields["metric_value"] = entry.metric_value
    if user_id is not None:
        fields["user_id"] = user_id

    client_logger = get_client_logger()
    message = _truncate(entry.message, MAX_MESSAGE_LENGTH) or ""
    level = (entry.level or "INFO").upper()
    if level == "ERROR":
        client_logger.error(message, extra={"client_fields": fields})
    elif level == "WARN":
        client_logger.warning(message, extra={"client_fields": fields})
    else:
        client_logger.info(message, extra={"client_fields": fields})


@router.post("/client", summary="前端日志批量上报")
async def collect_client_logs(
    body: ClientLogBatch = Body(...),
    user: Optional[UserContext] = Depends(get_current_user_optional),
):
    try:
        user_id = user.id if user else None
        for entry in body.logs:
            _write_entry(entry, user_id)
    except Exception:  # noqa: BLE001 - 落盘失败不应影响主流程，仅记录后返回失败
        logger.exception("前端日志落盘失败")
        return error("日志写入失败", ResultCode.SYSTEM_EXECUTION_ERROR.code)
    return success()
